src/ptychography/reconstruction/ssb/udf.py

```diff
@@ -1,11 +1,66 @@
 import numpy as np
+import numba
 
 from libertem.udf import UDF
 from libertem.common.container import MaskContainer
 
 from ptychography.reconstruction.ssb.trotters import generate_masks
 
 
+@numba.njit(fastmath=True, cache=True, parallel=True)
+def rmatmul_csc_fourier(n_threads, left_dense, right_data, right_indices, right_indptr,
+                        coordinates, row_exp, col_exp, res_inout):
+    '''
+    Fold :meth:`SSB_UDF.merge_dot_result` into a sparse dot product from
+    :meth:`libertem.common.numba._rmatmul_csc`
+
+    The result can be directly merged into the result buffer instead of
+    instantiating the intermediate dot result, which can be large. That way,
+    this method can process entire memory-mapped partitions efficiently.
+    Furthermore, it allows early skipping of trotters that are empty in this tile.
+
+    It uses multithreading to process different parts of a tile in parallel.
+    '''
+    left_rows = left_dense.shape[0]
+    p_size = col_exp.shape[1]
+    q_size = row_exp.shape[1]
+    # We subdivide in blocks per thread so that each thread
+    # writes exclusively to its own part of an intermediate result buffer.
+    # Using prange and automated '+=' merge leads to wrong results when using threading.
```

Comment:
Yeah, this is one of the caveats: `+=` is not thread safe. `TypedList.append` has the same problem, to reducing into a `(n_threads, ...)` array is the correct solution, I agree.

Reply:
In the Numba documentation they mention support for `+=` merge operations with `prange()`: https://numba.pydata.org/numba-doc/latest/user/parallel.html#explicit-parallel-loops

I am not sure why it didn't work here. In any case, per-thread intermediate buffers work just the same and leave no room for speculation. :-) Also, nice that `UDF.meta.threads_per_worker` was immediately useful, although one could also use `numba.get_num_threads()`.

Reply:
Ah, this might be it: "Care should be taken, however, when reducing into slices or elements of an array if the elements specified by the slice or index are written to simultaneously by multiple parallel threads. The compiler may not detect such cases and then a race condition would occur."